### PR TITLE
Networking: added support for PPP <-> Ethernet gateway

### PR DESCRIPTION
**Pull request by tridge**
If a board is configured with both PPP and ethernet then it gets a new option to be a PPP <-> Ethernet gateway. This can be done with either a normal or peripheral firmware by setting NET_OPTIONS=1

The IP address of the local end of the PPP link is set to be the same as the ethernet IP address. The IP address of the remote end of the link can be set with NET_REMPPP_IP or if that is left as 0 (ie. 0.0.0.0) then it will use the local ethernet address + 1.

If the remote IP address on the PPP link is on the same subnet as the ethernet interface then proxyarp is automatically enabled so that hosts on the ethernet network can connect without needing a route added.

This PR also adds a Pixhawk6X-PPPGW periph target that demonstrates how to build a peripheral which acts as a PPP gateway. The peripheral can be configured over CAN and will bridge the PPP and ethernet interfaces.

### CODE REVIEW

File: Tools/AP_Periph/networking.cpp [hunk 1]
```diff
@@ -137,12 +137,32 @@ const AP_Param::GroupInfo Networking_Periph::var_info[] {
...
+#if AP_NETWORKING_BACKEND_PPP
+    // @Param: PPP_PORT
+    // @DisplayName: PPP serial port
+    // @Description: PPP serial port
+    // @Range: -1 10
+    AP_GROUPINFO("PPP_PORT", 20, Networking_Periph, ppp_port, AP_PERIPH_NET_PPP_PORT_DEFAULT),
+
+    // @Param: PPP_BAUD
+    // @DisplayName: PPP serial baudrate
+    // @Description: PPP serial baro.cpp
```
Comment (magicrub) [inline Tools/AP_Periph/networking.cpp:149]: copy/paste bug `baro.cpp`

File: Tools/AP_Periph/networking.h [hunk 2]
```diff
@@ -54,7 +54,20 @@ class Networking_Periph {
...
 
     AP_Networking networking_lib;
+
+#if AP_NETWORKING_BACKEND_PPP
+    AP_Int8 ppp_port;
+    AP_Int32 ppp_baud;
+#endif
 };
 
+#ifndef AP_PERIPH_NET_PPP_PORT_DEFAULT
```
Comment (magicrub) [inline Tools/AP_Periph/networking.h:64]: why aren't these at the top?

File: libraries/AP_Networking/AP_Networking.cpp [hunk 3]
```diff
@@ -86,6 +86,20 @@ const AP_Param::GroupInfo AP_Networking::var_info[] = {
     AP_SUBGROUPINFO(param.test_ipaddr, "TEST_IP", 8,  AP_Networking, AP_Networking_IPV4),
 #endif
 
+    // @Param: OPTIONS
+    // @DisplayName: Networking options
+    // @Description: Networking options
+    // @Bitmask: 0:EnablePPP ethernet gateway
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking.cpp:92]: Ethernet should be capitalized

File: libraries/AP_Networking/AP_Networking.cpp [hunk 4]
```diff
@@ -129,8 +143,19 @@ void AP_Networking::init()
     }
 #endif
 
+#if AP_NETWORKING_PPP_GATEWAY_ENABLED
+    if (option_is_set(OPTION::PPP_ETHERNET_GATEWAY)) {
+        /*
+          when we are a PPP/Ethernet gateway we bring up the ethernet first
+         */
+        backend = new AP_Networking_ChibiOS(*this);
+        backend2 = new AP_Networking_PPP(*this);
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking.cpp:152]: wouldn't backend_PPP be a better name?

**Review by magicrub — CHANGES_REQUESTED** [review]

File: Tools/AP_Periph/networking.cpp (hunk 1 above)
Comment (tridge) [reply Tools/AP_Periph/networking.cpp:149]: oops! was auto-complete bug :-)

File: Tools/AP_Periph/networking.h (hunk 2 above)
Comment (tridge) [reply Tools/AP_Periph/networking.h:64]: you mean top of the class in header? any particular reason?

File: libraries/AP_Networking/AP_Networking.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking.cpp:92]: hah, I never knew it was supposed to be capitalised
https://english.stackexchange.com/questions/315198/does-ethernet-need-to-be-capitalised
changed

File: libraries/AP_Networking/AP_Networking.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_Networking/AP_Networking.cpp]: changed

File: Tools/AP_Periph/networking.h (hunk 2 above)
Comment (magicrub) [reply Tools/AP_Periph/networking.h:64]: because all other ifndef's are at the top everywhere else!

**Review by magicrub — APPROVED** [review]
you still have that ifndef in a weird place but I won't hold up the merge for that

File: Tools/AP_Periph/networking.h (hunk 2 above)
Comment (tridge) [reply Tools/AP_Periph/networking.h]: ahh, I see what you mean, fixed

**Comment by tridge** [comment]
> you still have that ifndef in a weird place but I won't hold up the merge for that

fixed it

**Pull request merged**